day1/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+const INITIAL_FLOOR: i32 = 0;
+
+/// Updates the floor Santa would reach after following the passed `instruction`.
+/// Returns `Some(new_floor)` to be compatible with [`std::iter::Iterator::scan`].
+fn process_floor(current_floor: &mut i32, instruction: char) -> Option<i32> {
+    match instruction {
+        '(' => *current_floor += 1,
+        ')' => *current_floor -= 1,
+        _ => panic!("invalid input"),
+    }
+    Some(*current_floor)
+}
+
+/// Returns the last floor Santa would be on after following all the instructions in `input`.
+pub fn part1(input: &str) -> i32 {
+    input
+        .chars()
+        .scan(INITIAL_FLOOR, process_floor)
+        .last()
+        .unwrap()
+}
+
+/// Returns the first day Santa be enter the basement on while following all the instructions in
+/// `input`.
+pub fn part2(input: &str) -> usize {
+    input
+        .chars()
+        .scan(INITIAL_FLOOR, process_floor)
+        .zip(1..)
+        .find(|&(floor, _)| floor < 0)
+        .unwrap()
+        .1
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I pulled this into `lib.rs` so I could add tests. I realize now I didn't need to do that. But now I can write benchmarks which showed these results when switching to `scan`:

```
part1                   time:   [17.598 us 17.675 us 17.775 us]
                        change: [-21.286% -20.482% -19.649%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild
  5 (5.00%) high severe

part2                   time:   [2.7329 us 2.7529 us 2.7792 us]
                        change: [+1.8647% +3.2406% +4.3881%] (p = 0.00 < 0.05)
                        Performance has regressed.
```
🤷 Interested in chatting about those numbers.